CARDS-2755: YE Outpatient survey

marta- opened the pull request:
Four of the five subtasks are tackled so far on this branch:
* [CARDS-2756](https://phenotips.atlassian.net/browse/CARDS-2756): _Outpatient survey: questionnaire definition_
* [CARDS-2757](https://phenotips.atlassian.net/browse/CARDS-2757): _Outpatient survey: ClinicMapping / Survey setup_
* [CARDS-2759](https://phenotips.atlassian.net/browse/CARDS-2759): _Outpatient survey email setup_
  * Email templates, sender / subject configuration, and email schedule (2 days and 4 days)
* [CARDS-2760](https://phenotips.atlassian.net/browse/CARDS-2760): _Outpatient survey responses export_

[CARDS-2756]: https://phenotips.atlassian.net/browse/CARDS-2756?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-2757]: https://phenotips.atlassian.net/browse/CARDS-2757?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-2759]: https://phenotips.atlassian.net/browse/CARDS-2759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-2760]: https://phenotips.atlassian.net/browse/CARDS-2760?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Co-authored-by: veronikaslc <veronika.koltunov@gmail.com>

marta- committed:
CARDS-2755 / CARDS-2727: Outpatient survey - ClinicMapping / Survey setup

marta- committed:
CARDS-2755 / CARDS-2759: Outpatient survey email setup

Added email templates, subject + sender configuration, and email schedule configuration.

Not tested.

marta- committed:
CARDS-2755 / CARDS-2760: Outpatient survey responses export configuration

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Updated the freetext question description, to match the one currently used in existing suveys (e.g. Inpatient).

marta- committed:
CARDS-2755 / CARDS-2759: Outpatient survey email setup
- Fix email ${unit} and ${date} not being interpreted in the email text

acrowthe committed:
CARDS-2755 / CARDS-2759: Outpatient survey email setup
- Fix email templates not being loaded into CARDS

Hajany commented:
Can we add the visit date to the landing page as well? Exactly how it is in the email
".... recent outpatient visit at TG-THROMBOSIS AND HEMOSTASIS on Sept 23, 2025."

<img width="1030" height="904" alt="image" src="https://github.com/user-attachments/assets/67665abb-9d94-4061-ae10-4a7f3fcf9f60" />

Hajany commented:
The survey link expiration date is off, this survey is for a visit on Sept 23, 2025 

- Initial email sent Sept 25, 2025
- Reminder email sent Sept 27, 2025
- Link should remain active until Oct 11, 2025
- Today's date is Sept 29, 2025 and it says link is active for another 24 days (when it should only be active for another 12 days)

<img width="1257" height="897" alt="image" src="https://github.com/user-attachments/assets/e8bbe401-fc6b-42d7-bf1c-a7a3a106dc5b" />

marta- committed:
CARDS-2755 / CARDS-2727: Outpatient survey - ClinicMapping / Survey setup

Update number of days after the visit while survey is valid

marta- commented:
> Can we add the visit date to the landing page as well? Exactly how it is in the email
> ".... recent outpatient visit at TG-THROMBOSIS AND HEMOSTASIS on Sept 23, 2025."

This requires code changes to be able to accommodate multivalued fields (the answer that holds TG-THROMBOSIS AND HEMOSTASIS is `provider`, which is multivalued), and formatted dates. Will be done shortly in a separate PR.

marta- committed:
CARDS-2755 / CARDS-2727: Outpatient survey - ClinicMapping / Survey setup

Displayed the clinic and date instead of the name of the hospital on the landing page.
To support this change, removed the access restrictions to the `provider` question and answer from Visit information.

Hajany commented:
Pulling Provider field instead of Location now
<img width="972" height="903" alt="image" src="https://github.com/user-attachments/assets/079fc79d-d709-4137-bc1b-3e618df9b6b6" />
<img width="866" height="907" alt="image" src="https://github.com/user-attachments/assets/d98af3bd-01e3-482f-9a12-f21c75163310" />

marta- commented:
> Pulling Provider field instead of Location now

That is the correct behavior. In YE, we store the unit name in the `provider` field and the hospital name in the `location` field. See for example the clarity mapping for the PM Outpatient Oncology survey: https://github.com/data-team-uhn/cards/blob/dev/prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clarityImport/PMH-OO.xml#L179-L192 . The UHN Outpatient Care survey clarity mapping is not implemented yet, but the behavior will be similar, with the visited department stored in the `provider` field.

acrowthe commented:
Found a weird bug with the outpatient questionnaire:
- Create a new OC form
- Fill out page 1 with `Telephone`
- Proceed to the next page
- Fill out page 2 with `No technical problems` and `Yes`
- Proceed to the next page
- Refresh the page, or otherwise close and re-open the form
- Observe that you are on page 1, with `Telephone` selected, as expected
- Click next
Expected result:
- page 2 opens
Actual result:
- page 2 is skipped and page 3 opens

This bug occurs in both the Clinician UI and the patient portal

I was not able to reproduce this on dev with a simple conditional test form but I have not narrowed down what is causing OC page 2 to be skipped on subsequent form edits. I'm not sure if there's an issue somewhere in the conditions for this form or if there's a bug in conditional processing.

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Made question 1 (visit type) a text instead of long in order to use human-readable values and  work around a conditional evaluation issue

marta- commented:
> Found a weird bug with the outpatient questionnaire:

Addressed in c621b5a .

marta- commented:
> Found a weird bug with the outpatient questionnaire:

Addressed in e3f6af6.

Root cause reported as https://phenotips.atlassian.net/browse/CARDS-2799 and addressed in #2111.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/ReminderNotification/bodyTemplate.html:
```diff
@@ -0,0 +1,64 @@
+      receiving this email because our records show that you had a recent
+      outpatient clinic visit at ${unit} on ${date},
+      at the University Health Network (UHN).
+	  Our records indicate that we sent you an optional survey asking about
```
The initial email has a paragraph break here. Is it intentionally missing from the reminder?

marta- replied:
There's no break in the specs of the Reminder https://phenotips.atlassian.net/browse/CARDS-2759.

I have no strong opinion either way.

Hajany replied:
I also have no strong opinion either way but to keep it simple I say we follow the specs

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/ReminderNotification/bodyTemplate.txt:
```diff
@@ -0,0 +1,48 @@
+receiving this email because our records show that you had a recent
+outpatient clinic visit at ${unit} on ${date},
+at the University Health Network (UHN).
+Our records indicate that we sent you an optional survey asking about
```
Same here, break missing in the reminder.

marta- replied:
See previous comment.

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/ReminderNotification/bodyTemplate.html:
```diff
@@ -0,0 +1,64 @@
+    </p>
+    <p>
+      <strong>The survey will be available until ${expirationDate}. Please do not delay.</strong>
+    </p>
```
We say this twice, see 2 paras above.

marta- replied:
To be removed:

```suggestion
```

sdumitriu commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/ReminderNotification/bodyTemplate.txt:
```diff
@@ -0,0 +1,48 @@
+the quality of care patients receive. Only a summary of the overall
+responses will be shared to help improve care.
+
+If you have any questions or concerns or wish to be removed from the survey,
```
The wording in the HTML is a bit different:


```suggestion
If you have any questions, concerns, or wish to be removed from the survey,
```

marta- replied:
For both the initial email and the reminder the specs say "If you have any questions or concerns or wish to be removed ...". I like the version with commas more. We could change it everywhere if it's not a big deal to the client @Hajany .

Hajany replied:
I'm ok with the commas, let's change this

sdumitriu committed:
CARDS-2755 / CARDS-2727: Outpatient survey - ClinicMapping / Survey setup
Add CopyAnswers configuration

sdumitriu committed:
CARDS-2755 / CARDS-2760: Outpatient survey responses export configuration
Also include the OCPE survey in the raw and daily exports

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Updated to final version of the survey
- minor reordering of answer options in Module 1
- 3 new questions at the beginning of Module 2

marta- committed:
CARDS-2755 / CARDS-2802: Outpatient landing page setup

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/InitialInvitation/bodyTemplate.html:
```diff
@@ -0,0 +1,65 @@
+    </p>
+    <p>
+      <strong>The survey will be available until ${expirationDate}. Please do not delay.</strong>
+    </p>
```
To be removed:

```suggestion
```

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clinics/UHN-OC/mailTemplates/InitialInvitation/bodyTemplate.txt:
```diff
@@ -0,0 +1,48 @@
+the quality of care patients receive. Only a summary of the overall
+responses will be shared to help improve care.
+
+If you have any questions or concerns or wish to be removed from the survey,
```
```suggestion
If you have any questions, concerns, or wish to be removed from the survey,
```

marta- committed:
CARDS-2755 / CARDS-2759: Outpatient survey email setup

Removed redundant expiration date information.

marta- committed:
CARDS-2755 / CARDS-2759: Outpatient survey email setup

Ensure consistent phrasing across email templates.

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Use non-breaking spaces in the question matrix answer option labels

Hajany commented:
Questions 3 and 4 of the Virtual Care section are pre populated if you go back and change answer to Question 1, should clear out. Also red box indicating mandatory pops up when not needed yet
[screen-capture (16).webm](https://github.com/user-attachments/assets/3605bbfd-8328-4c25-87f9-dc6dbf428edb)

Hajany commented:
Missing "Not applicable" answer option at the end for this question
<img width="885" height="311" alt="image" src="https://github.com/user-attachments/assets/b5dabfb5-b35f-4d6d-80b2-778636258b7d" />
Specs
<img width="643" height="165" alt="image" src="https://github.com/user-attachments/assets/ded5d327-29e3-49d2-8fe6-ca4252242d92" />

marta- committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition

Add missing NA option to question 21.

sdumitriu committed:
CARDS-2755 / CARDS-2756: Outpatient survey - questionnaire definition
Emphasize the most important part of the free text question description

sdumitriu committed:
[misc] Replace tab with spaces
